pkg/didcomm/common/service/connection_record.go

```diff
@@ -0,0 +1,205 @@
+/*
+ *
+ * Copyright SecureKey Technologies Inc. All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * /
+ *
+ */
+
+package service
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/hyperledger/aries-framework-go/pkg/storage"
+)
+
+const (
+	nameSpace          = "didexchange"
+	keyPattern         = "%s_%s"
+	connIDKeyPrefix    = "conn"
+	connStateKeyPrefix = "connstate"
+	// limitPattern with `~` at the end for lte of given prefix (less than or equal)
+	limitPattern    = "%s~"
+	keySeparator    = "_"
+	stateIDEmptyErr = "stateID can't be empty"
+)
+
+// KeyPrefix is prefix builder for storage keys
+type KeyPrefix func(...string) string
+
+type provider interface {
+	TransientStorageProvider() storage.Provider
+	StorageProvider() storage.Provider
+}
+
+// ConnectionRecord contain info about did exchange connection
+type ConnectionRecord struct {
+	ConnectionID    string
+	State           string
+	ThreadID        string
+	TheirLabel      string
+	TheirDID        string
+	MyDID           string
+	ServiceEndPoint string
+	RecipientKeys   []string
+	InvitationID    string
+	InvitationDID   string
+	Implicit        bool
+	Namespace       string
+}
+
+// NewConnectionRecorder returns new connection recorder instance
+func NewConnectionRecorder(p provider) (*ConnectionRecorder, error) {
+	store, err := p.StorageProvider().OpenStore(nameSpace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new connection recorder : %w", err)
+	}
+
+	transientStore, err := p.TransientStorageProvider().OpenStore(nameSpace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new connection recorder : %w", err)
+	}
+
+	return &ConnectionRecorder{transientStore: transientStore, store: store}, nil
+}
+
+// ConnectionRecorder takes care of connection related persistence features
+type ConnectionRecorder struct {
+	transientStore storage.Store
+	store          storage.Store
+}
+
+// GetConnectionRecord return connection record based on the connection ID
+func (c *ConnectionRecorder) GetConnectionRecord(connectionID string) (*ConnectionRecord, error) {
+	rec, err := getAndUnmarshal(GetConnectionKeyPrefix()(connectionID), c.store)
+	if err != nil {
+		if errors.Is(err, storage.ErrDataNotFound) {
+			return getAndUnmarshal(GetConnectionKeyPrefix()(connectionID), c.transientStore)
+		}
+
+		return nil, err
+	}
+
+	return rec, nil
+}
```

Comment:
looks better
```
func (c *ConnectionRecorder) GetConnectionRecord(connectionID string) (*ConnectionRecord, error) {
   rec, err := getAndUnmarshal(GetConnectionKeyPrefix()(connectionID), c.store)
    if errors.Is(err, storage.ErrDataNotFound) {
		return getAndUnmarshal(GetConnectionKeyPrefix()(connectionID), c.transientStore)
	}

	return rec, err
}
```

Reply:
will make sure it doesn't panic when err = nil

Reply:
done